From 184e25ebc2cffc5ad8f3c5d5d8a6f8428f1d120e Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Wed, 22 Mar 2017 14:07:30 -0700
Subject: [PATCH] MANTA-3198 enable pinger on muskie -> authcache and muskie ->
 mako connections Reviewed by: David Pacheco <dap@joyent.com> Reviewed by:
 Jordan Hendricks <jordan.hendricks@joyent.com> Approved by: Jordan Hendricks
 <jordan.hendricks@joyent.com>

---
 etc/config.coal.json           |  3 +++
 main.js                        | 13 ++++++++++++-
 package.json                   |  2 +-
 sapi_manifests/muskie/template | 18 ++++++++++++++++++
 4 files changed, 34 insertions(+), 2 deletions(-)

diff --git a/etc/config.coal.json b/etc/config.coal.json
index 92c9bdd..76947d7 100644
--- a/etc/config.coal.json
+++ b/etc/config.coal.json
@@ -43,6 +43,8 @@
         "initialDomains": [
           "authservice.coal.joyent.us"
         ],
+        "pingInterval": 60000,
+        "tcpKeepAliveInitialDelay": 10000,
         "spares": 8,
         "maximum": 200,
         "recovery": {
@@ -90,6 +92,7 @@
         "delay": 250,
         "maxIdleTime": 10000,
         "maxClients": 50,
+        "pingInterval": 14400000,
         "retry": {
           "retries": 2
         },
diff --git a/main.js b/main.js
index f6933c8..baef401 100644
--- a/main.js
+++ b/main.js
@@ -215,10 +215,17 @@ function createCueballHttpAgent(cfg) {
     /* Used only for connections to sharks. */
     var sharkCueball = {
         resolvers: sharkCfg.resolvers,
+
         spares: sharkCfg.spares,
         maximum: sharkCfg.maximum,
-        linger: sharkCfg.maxIdleTime,
+        /*
+         * Note that this path doesn't actually have to be handled by the
+         * authcache (any non-5xx response code is accepted, e.g. 404 is fine).
+         */
+        ping: '/ping',
+        pingInterval: sharkCfg.pingInterval,
         tcpKeepAliveInitialDelay: sharkCfg.maxIdleTime,
+
         log: sharkCfg.log,
         recovery: {
             default: {
@@ -227,6 +234,10 @@ function createCueballHttpAgent(cfg) {
                 maxTimeout: sharkCfg.maxTimeout,
                 delay: sharkCfg.delay
             },
+            /*
+             * Avoid SRV retries, since authcache doesn't currently register
+             * any useable SRV records for HTTP (it only registers redis)
+             */
             'dns_srv': {
                 retries: 0,
                 timeout: sharkCfg.connectTimeout,
diff --git a/package.json b/package.json
index bb7cf46..a743a67 100644
--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
         "backoff": "2.3.0",
         "bunyan": "0.22.1",
         "bunyan-syslog": "0.2.2",
-        "cueball": "2.2.0",
+        "cueball": "2.2.3",
         "dashdash": "1.3.2",
         "deep-equal": "0.0.0",
         "dtrace-provider": "0.2.8",
diff --git a/sapi_manifests/muskie/template b/sapi_manifests/muskie/template
index 61d6f7e..f40f604 100644
--- a/sapi_manifests/muskie/template
+++ b/sapi_manifests/muskie/template
@@ -53,6 +53,21 @@
       "{{AUTH_SERVICE}}"
     ],
 
+    {{!
+      Make sure that if a socket is left idle for >60sec, a dummy "ping"
+      request is made on it. This stops the node.js 2-minute server socket
+      timeout on the authcache side from closing our sockets (and potentially
+      racing with our use of that socket).
+    }}
+    "pingInterval": 60000,
+
+    {{!
+      Separately, we want to enable TCP-level keep-alives on the sockets, to
+      ensure that we notice quickly if an entire CN running an authcache panics
+      or is netsplit from us, to avoid handing it requests that will fail.
+    }}
+    "tcpKeepAliveInitialDelay": 10000,
+
     {{!
       The spares value here should be larger than sharkConfig.spares, by
       a factor probably >2, <10. Mahi connections are shared amongst all reqs
@@ -118,6 +133,9 @@
     "maxIdleTime": 10000,
     "maxClients": 50,
 
+    {{! nginx has much longer max idle times than node.js }}
+    "pingInterval": 14400000,
+
     {{!
       We want spares to be small: for the sake of prudence we don't want lots
       of idle sockets sitting around for no reason. We also want it to be large:
-- 
2.21.0

